internal/slack/files.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	// Open the input archive.
+	r, err := zip.OpenReader(inputArchive)
+	if err != nil {
+		return fmt.Errorf("could not open input archive for reading: %s\n", inputArchive)
+	}
+	defer r.Close()
+
+	// Open the output archive.
+	f, err := os.Create(outputArchive)
+	if err != nil {
+		return fmt.Errorf("could not open the output archive for writing: %s\n\n%s", outputArchive, err)
+	}
+	defer f.Close()
+
+	// Create a zip writer on the output archive.
+	w := zip.NewWriter(f)
+
+	// Run through all the files in the input archive.
+	for _, file := range r.File {
+
+		// Open the file from the input archive.
+		inReader, err := file.Open()
+		if err != nil {
+			fmt.Printf("Failed to open file in input archive: %s\n\n%s", file.Name, err)
+			continue
+		}
+
+		// Read the file into a byte array.
+		inBuf, err := ioutil.ReadAll(inReader)
+		if err != nil {
+			fmt.Printf("Failed to read file in input archive: %s\n\n%s", file.Name, err)
+			continue
+		}
+
+		// Now write this file to the output archive.
+		outFile, err := w.Create(file.Name)
+		if err != nil {
+			return fmt.Errorf("Failed to create file in output archive: %s\n\n%s", file.Name, err)
+		}
+		_, err = outFile.Write(inBuf)
+		if err != nil {
+			fmt.Printf("Failed to write file in output archive: %s\n\n%s", file.Name, err)
+			continue
+		}
+
+		// Check if the file name matches the pattern for files we need to parse.
+		splits := strings.Split(file.Name, "/")
+		if len(splits) == 2 && !strings.HasPrefix(splits[0], "__") && strings.HasSuffix(splits[1], ".json") {
+			// Parse this file.
+			err = processChannelFile(w, file, inBuf)
+			if err != nil {
+				fmt.Printf("%s", err)
+				continue
+			}
+		}
+	}
+
+	// Close the output zip writer.
+	err = w.Close()
+	if err != nil {
+		fmt.Printf("Failed to close the output archive.\n\n%s", err)
+	}
+
+	return nil
+}
+
+// processChannelFile actually fetches and adds a found file to the
+// archive specified at file
+func processChannelFile(w *zip.Writer, file *zip.File, inBuf []byte) error {
+	// Parse the JSON of the file.
+	var posts []SlackPost
+	if err := json.Unmarshal(inBuf, &posts); err != nil {
+		return errors.New("Couldn't parse the JSON file: " + file.Name + "\n\n" + err.Error() + "\n")
+	}
+
+	// Loop through all the posts.
+	for _, post := range posts {
+		// Support for legacy file_share posts.
+		if post.Subtype == "file_share" {
+			// Check there's a File property.
+			if post.File == nil {
+				log.Print("file_share post has no File property: " + post.Ts + "\n")
+				continue
+			}
+
+			// Add the file as a single item in the array of the post's files.
+			post.Files = []*SlackFile{post.File}
+		}
+
+		// If the post doesn't contain any files, move on.
+		if post.Files == nil {
+			continue
+		}
+
+		// Loop through all the files.
+		for _, file := range post.Files {
+			// Check there's an Id, Name and either UrlPrivateDownload or UrlPrivate property.
```

Comment:
nit: I would consider extracting logic executed inside loop to separate function.